linux: fix uv_available_parallelism using cgroup

waltoss opened the pull request:
PR to fix https://github.com/libuv/libuv/issues/4146

bnoordhuis commented on src/unix/core.c:
```diff
@@ -1739,25 +1739,40 @@ int uv__search_path(const char* prog, char* buf, size_t* buflen) {
+
+    uv__cpu_constraint c = uv__get_constrained_cpu();
+    if (c.period_length > 0) {
+        int rc_with_cgroup = (int)(c.quota_per_period /
+                                   c.period_length * c.proportions);
+        if (rc_with_cgroup < rc) {
+            rc = rc_with_cgroup;
+        }
```
Style: var declarations go at the top; assignment separate from declaration; no braces around single if-statements

Substance: is the implicit cast to double intentional? `long long` can hold values >= 2**53, more than what double can accurately represent.

bnoordhuis commented on src/unix/linux.c:
```diff
@@ -2281,6 +2281,131 @@ uint64_t uv_get_available_memory(void) {
 
+#define BUF_SIZE 1024
+
+static uv__cpu_constraint uv__get_cgroupv2_constrained_cpu(const char *cgroup) {
```
Style: star leans left
```suggestion
static uv__cpu_constraint uv__get_cgroupv2_constrained_cpu(const char* cgroup) {
```

bnoordhuis commented on src/unix/linux.c:
```diff
@@ -2281,6 +2281,131 @@ uint64_t uv_get_available_memory(void) {
+  int cgroup_size;
+  char *cgroup_trimmed;
+
+  // Trim ending \n by replacing it with a 0
```
Style: c89 comments
```suggestion
  /* Trim ending \n by replacing it with a 0 */
```

bnoordhuis commented on src/unix/linux.c:
```diff
@@ -2281,6 +2281,131 @@ uint64_t uv_get_available_memory(void) {
+  char *cgroup_trimmed;
+
+  // Trim ending \n by replacing it with a 0
+  cgroup_trimmed = cgroup + strlen("0::/");          // Skip the prefix "0::/"
```
It's not wrong and with most compilers it amounts to the same generated machine code but idiomatic libuv source code uses sizeof for string literals:
```suggestion
  cgroup_trimmed = cgroup + sizeof("0::/") - 1;          // Skip the prefix "0::/"
```

bnoordhuis commented on src/unix/linux.c:
```diff
@@ -2281,6 +2281,131 @@ uint64_t uv_get_available_memory(void) {
+  cgroup_size = (int)strcspn(cgroup_trimmed, "\n");  // Find the first slash
+
+  // Construct the path to the cpu.max file
+  snprintf(path, sizeof(path), "/sys/fs/cgroup/%.*s/cpu.max", cgroup_size,
+            cgroup_trimmed);
```
```suggestion
  snprintf(path, sizeof(path), "/sys/fs/cgroup/%.*s/cpu.max", cgroup_size,
           cgroup_trimmed);
```

bnoordhuis commented on src/unix/linux.c:
```diff
@@ -2281,6 +2281,131 @@ uint64_t uv_get_available_memory(void) {
+            cgroup_trimmed);
+
+  // Read cpu.max
+  if (uv__slurp(path, buf, BUF_SIZE) >= 0) {
```
```suggestion
  if (uv__slurp(path, buf, sizeof(buf)) >= 0) {
```

bnoordhuis commented on src/unix/linux.c:
```diff
@@ -2281,6 +2281,131 @@ uint64_t uv_get_available_memory(void) {
+                                             int *cgroup_size) {
+  char *cgroup_cpu;
+
+  /* Seek to the cpu controller line. */
+  cgroup_cpu = strchr(cgroup, ':');
+  while (cgroup_cpu != NULL && strncmp(cgroup_cpu, ":cpu,", 5)) {
+      cgroup_cpu = strchr(cgroup_cpu, '\n');
+      if (cgroup_cpu != NULL) cgroup_cpu = strchr(cgroup_cpu, ':');
+  }
```
Maybe I'm missing something but aren't you basically doing `cgroup_cpu = strstr(cgroup_cpu, ":cpu,")` here?

bnoordhuis commented on src/unix/linux.c:
```diff
@@ -2281,6 +2281,131 @@ uint64_t uv_get_available_memory(void) {
+  char controllers[256];
+
+  // Read the cgroup from /proc/self/cgroup
+  strcpy(path, "/proc/self/cgroup");
```
Verboten :) - just always use snprintf()

bnoordhuis reviewed (commented):
Thanks for the PR. I pointed out single instances of style issues but please fix them all.

The ASAN buildbot is known to be problematic at the moment. You can just ignore it.

waltoss commented on include/uv.h:
```diff
@@ -1347,6 +1347,13 @@ UV_EXTERN uv_pid_t uv_os_getppid(void);
+  long long quota_per_period;
+  long long period_length;
+  double proportions;
+} uv__cpu_constraint;
```
How do I include properly while being able to test it in test-platform-output.c ?

bnoordhuis replied:
Take a look at e.g. test/test-idna.c, we do that by #including source files directly. You may need to split it off into a separate source file in order for that to work; in that case remember to update both CMakeLists.txt and Makefile.am.

waltoss replied:
I have different leads to implement the test
- Either refactor `uv__get_constrained_cpu` into a separate `src/linux-cgroup.c` for instance, but my issue is my code depends on `uv__slurp` so I would need to extract it as well. Would it make sense ?
- I tried to include `internal.h` or `uv-common.h` in the test but the linker does not find `uv__get_constrained_cpu`, like src/unix/linux.c is not link for building test, am I wrong ?
- Third option is to drop the test

I am no C expert, so any help is welcome

waltoss replied:
What about making `uv_get_constrained_cpu` an external definition just like `uv_get_constrained_memory`. It would simplify the tests + align with the memory implementation of cgroups

I would suggest the following documentation

```rst
.. c:function:: uv_cpu_constraint uv_get_constrained_cpu(void)

    Retrieves the CPU constraints imposed on the process by the operating system
    or container management system (e.g., cgroups on Linux). This function
    returns a `uv_cpu_constraint` structure that contains the details of the
    CPU constraints.

    If there are no CPU constraints, or if the constraints are unknown, the
    fields of the returned structure are set to zero. If there is a constraining
    mechanism but no specific constraints are set, the relevant fields are set
    to their maximum possible values.

    .. versionadded:: XXXX

.. c:type:: uv_cpu_constraint

    A structure representing the CPU constraints of a process.

    :c:member:`long long quota_per_period`
        The maximum amount of CPU time (in microseconds) that the process can
        consume in a given period. If the value is `0`, there is no quota
        enforced. If the value is `LLONG_MAX`, the quota is effectively
        unlimited.

    :c:member:`long long period_length`
        The length of the period (in microseconds) for which the quota is
        measured. A shorter period with the same quota allows for more
        fine-grained control of the process's CPU usage.

    :c:member:`double proportions`
        The weight or proportion of CPU resources available to the process
        relative to other processes. This is a normalized value, typically
        between `0.0` and `1.0`, where `1.0` represents full access to a single
        CPU core, and `0.5` would represent half of a CPU core. This value is
        used in conjunction with `quota_per_period` and `period_length` to
        determine the actual CPU time available to the process.
```

And update `uv_available_parallelism` documentation accordingly

What do you think ?

bnoordhuis replied:
I don't want to extend the public API on a whim because chances are good that without careful discussion and deliberation, it gets overfitted to the thing at hand (i.e., cgroups.)

You can open an issue to discuss but consider it out of scope for this pull request for now.

waltoss commented:
@bnoordhuis I fixed styling issues and added tests directly in test-platform-output.c. It might be a bit ugly though, happy to have your feedback on it

To test it manually I built it with the following Dockerfile

```Dockerfile
# Use the official Ubuntu base image
FROM ubuntu:20.04

# Update the package list
RUN apt-get update

# Install cmake and other build dependencies
RUN DEBIAN_FRONTEND=noninteractive apt-get install -y cmake libtool autoconf build-essential

# Fix : for some reason cmake is looking for gmake instead of make
RUN ln -s /usr/bin/make /usr/bin/gmake

# Set the working directory inside the container
WORKDIR /workspace

# Create a user 'libuv' and give ownership of /workspace to 'libuv'
RUN useradd -m libuv && chown -R libuv:libuv /workspace

# Switch to the 'libuv' user
USER libuv

# Copy the current directory contents into the container at /workspace
COPY --chown=libuv:libuv ./ /workspace/

# Run the build steps as 'libuv' user
RUN mkdir -p build
RUN (cd build && cmake .. -DBUILD_TESTING=ON)
RUN cmake --build build

# By default, run a shell
CMD ["/bin/sh", "-c", "/workspace/build/uv_run_tests platform_output"]
```

And following commands

```bash
$ docker build . libuv-test-cgroup
$ docker run --cpus="2" libuv-test-cgroup | grep available 
```

It correctly outputs 2 available cpus

bnoordhuis commented on src/unix/linux.c:
```diff
@@ -2281,6 +2281,124 @@ uint64_t uv_get_available_memory(void) {
+  int cgroup_size;
+  const char* cgroup_trimmed;
+
+  /* Trim ending \n by replacing it with a 0 */
+  cgroup_trimmed = cgroup + sizeof("0::/") - 1;      /* Skip the prefix "0::/" */
+  cgroup_size = (int)strcspn(cgroup_trimmed, "\n");  /* Find the first slash */
```
This function assumes the buffer always starts with "0::/" but I don't see anything at the call site that guarantees that. You should check `if (strncmp(buf, "0::/", 4))` either here or (for consistency with existing code) there.

waltoss replied:
Well in case of cgroup v2, `/sys/fs/cgroup/cgroup.controllers` [always starts with `0::/`, ](https://docs.kernel.org/admin-guide/cgroup-v2.html)
<img width="819" alt="image" src="https://github.com/libuv/libuv/assets/31201229/9bb61a92-89c2-43f5-b3e1-43d64607d63a">

Do I need that extra check ?

bnoordhuis replied:
Always validate that input is as expected. Someone may bind-mount another file over /sys/fs/cgroup/cgroup.controllers

bnoordhuis commented on src/unix/linux.c:
```diff
@@ -2281,6 +2281,124 @@ uint64_t uv_get_available_memory(void) {
+           cgroup_trimmed);
+
+  /* Read cpu.max */
+  if (uv__slurp(path, buf, sizeof(buf)) >= 0) {
```
If you can't read cpu.max, there's no point in going on to read cpu.weight, right? Just bail out.

This function should probably return an error instead of a struct so the caller can properly distinguish between success and error. Returning a partially initialized success value and leaving it up to the caller to figure out what's what is not a good idea.

waltoss replied:
OK, should I add an extra property in `uv__cpu_constraint` to flag it's an error ?

bnoordhuis commented on src/unix/linux.c:
```diff
@@ -2281,6 +2281,124 @@ uint64_t uv_get_available_memory(void) {
+
+  /* Construct the path to the cpu.max file */
+  snprintf(path, sizeof(path), "/sys/fs/cgroup/%.*s/cpu.cfs_quota_us",
+           cgroup_size, cgroup_cpu);
```
Unsafe: cgroup_cpu can be NULL, cgroup_size can be uninitialized. You should handle the is-NULL case a few lines up.

bnoordhuis commented on src/unix/linux.c:
```diff
@@ -2281,6 +2281,124 @@ uint64_t uv_get_available_memory(void) {
+  snprintf(path, sizeof(path), "/sys/fs/cgroup/%.*s/cpu.cfs_quota_us",
+           cgroup_size, cgroup_cpu);
+
+  if (uv__slurp(path, buf, sizeof(buf)) >= 0) {
+    sscanf(buf, "%lld", &constraint.quota_per_period);
+  }
```
Style: no braces

Substance: bail out (like my other comment)

bnoordhuis commented on src/unix/linux.c:
```diff
@@ -2281,6 +2281,124 @@ uint64_t uv_get_available_memory(void) {
+  char controllers[256];
+
+  /* Read the cgroup from /proc/self/cgroup */
+  snprintf(path, sizeof(path), "/proc/self/cgroup");
+  if (uv__slurp(path, cgroup, sizeof(cgroup)) < 0) {
```
```suggestion
  if (uv__slurp("/proc/self/cgroup", cgroup, sizeof(cgroup)) < 0) {
```
Likewise below.

bnoordhuis commented on src/unix/linux.c:
```diff
@@ -2281,6 +2281,124 @@ uint64_t uv_get_available_memory(void) {
+
+  /* Check if the system is using cgroup v2 */
+  snprintf(path, sizeof(path), "/sys/fs/cgroup/cgroup.controllers");
+  if (uv__slurp(path, controllers, sizeof(controllers)) >= 0)
```
If you're not going to use the contents, just `access("/sys/fs/cgroup/cgroup.controllers", F_OK)`.

That said... what's the reason for using cgroup.controllers here? Everywhere else we look at the format of /proc/self/cgroup to figure out if we're dealing with cgroups v1 or v2.

bnoordhuis commented on src/win/util.c:
```diff
@@ -511,7 +511,6 @@ int uv_uptime(double* uptime) {
   return 0;
 }
 
-
```
Please undo.

bnoordhuis commented on test/test-platform-output.c:
```diff
@@ -84,12 +84,78 @@ TEST_IMPL(platform_output) {
   printf("  page faults: %llu\n", (unsigned long long) rusage.ru_majflt);
   printf("  maximum resident set size: %llu\n",
-         (unsigned long long) rusage.ru_maxrss);
+         (unsigned long long)rusage.ru_maxrss);
```
Please undo.

bnoordhuis commented on test/test-platform-output.c:
```diff
@@ -84,12 +84,78 @@ TEST_IMPL(platform_output) {
   ASSERT_GE(par, 1);
   printf("uv_available_parallelism: %u\n", par);
 
+  #ifdef __linux__
```
```suggestion
#ifdef __linux__
```
(don't indent preprocessor directives)

bnoordhuis commented on test/test-platform-output.c:
```diff
@@ -84,12 +84,78 @@ TEST_IMPL(platform_output) {
+  unsigned int cgroup_par = 0;
+
+  // Check if the system is using cgroup v2
+  file = fopen("/sys/fs/cgroup/cgroup.controllers", "r");
```
Use access(2) but note my other comment about cgroup.controllers.

bnoordhuis commented on test/test-platform-output.c:
```diff
@@ -84,12 +84,78 @@ TEST_IMPL(platform_output) {
+    // Parse cgroup v1 to deduce parallelism constraints
+    file = fopen("/sys/fs/cgroup/cpu,cpuacct/cpu.cfs_quota_us", "r");
+    if (file) {
+      if (getline(&line, &len, file) != -1) {
```
`line` is malloc'd by getline() but never freed, it looks like.

bnoordhuis commented on test/test-platform-output.c:
```diff
@@ -84,12 +84,78 @@ TEST_IMPL(platform_output) {
+    file = fopen("/sys/fs/cgroup/cpu,cpuacct/cpu.cfs_quota_us", "r");
+    if (file) {
+      if (getline(&line, &len, file) != -1) {
+        long long quota = atoll(line);
```
`long long` isn't exactly wrong but in libuv we generally use int64_t or uint64_t. The latter is probably more appropriate because quotas are never negative.

bnoordhuis commented on test/test-platform-output.c:
```diff
@@ -84,12 +84,78 @@ TEST_IMPL(platform_output) {
+      }
+      fclose(file);
+    }
+  }
```
This code can probably be simplified a little if you simply try to read the v2 files first, then v1 files if that fails. The manual parsing can likely be reduced by leaning into fscanf().

bnoordhuis commented on test/test-platform-output.c:
```diff
@@ -207,3 +273,4 @@ TEST_IMPL(platform_output) {
 
   return 0;
 }
+
```
Please undo.

waltoss commented on src/unix/core.c:
```diff
@@ -1744,19 +1744,28 @@ unsigned int uv_available_parallelism(void) {
-  if (rc < 1)
+  uv__cpu_constraint c = uv__get_constrained_cpu();
+  if (c.period_length > 0) {
+    rc_with_cgroup = (int)((double)(c.quota_per_period / c.period_length) * c.proportions);
```
You are right, actually casting to int is useless here since I assign `rc` to `rc_with_cgroup` only if `rc_with_cgroup < rc < LONG_MAX`

waltoss commented:
@bnoordhuis have you other feedbacks ?

bnoordhuis commented:
Can you rebase on top of the v1.x branch? It looks like you're a few commits behind.

waltoss commented:
> Can you rebase on top of the v1.x branch? It looks like you're a few commits behind.

✅

waltoss commented:
@bnoordhuis I rebased on v1.x

santigimeno commented on src/unix/linux.c:
```diff
@@ -2270,6 +2270,136 @@ uint64_t uv_get_available_memory(void) {
+  if (sscanf(buf, "%15s %llu", quota_buf, &constraint->period_length) != 2)
+    return UV_EINVAL;
+
+  if (strcmp(quota_buf, "max") == 0)
```
```suggestion
  if (strncmp(quota_buf, "max", 3) == 0)
```

santigimeno commented on src/unix/linux.c:
```diff
@@ -2270,6 +2270,136 @@ uint64_t uv_get_available_memory(void) {
+  const char* cgroup_trimmed;
+  char quota_buf[16];
+
+  if (strncmp(cgroup, "0::/", 4) != 0)
+    return UV_EINVAL;
```
I think we can remove this check as it was done before calling the function

bnoordhuis replied:
I believe I suggested doing that as an extra sanity check / forward-proofing.

santigimeno commented on src/unix/linux.c:
```diff
@@ -2270,6 +2270,136 @@ uint64_t uv_get_available_memory(void) {
+  if (strcmp(quota_buf, "max") == 0)
+    constraint->quota_per_period = LLONG_MAX;
+  else
+    constraint->quota_per_period = strtoll(quota_buf, NULL, 10);
```
Should we check for errors in the conversion?

waltoss replied:
Is this the correct way to handle errors ?

```c
  if (strncmp(quota_buf, "max", 3) == 0)
    constraint->quota_per_period = LLONG_MAX;
  else {
    char* endptr;
    errno = 0;
    constraint->quota_per_period = strtoll(quota_buf, &endptr, 10);
    if ((errno == ERANGE && (constraint->quota_per_period == LLONG_MAX || constraint->quota_per_period == LLONG_MIN)) || (errno != 0 && constraint->quota_per_period == 0)) {
        return UV_EINVAL; // overflow
    }
    if (*endptr != '\0')
      return UV_EINVAL; // conversion failed
  }
```

waltoss replied:
Or maybe simpler

```c
  if (strncmp(quota_buf, "max", 3) == 0)
    constraint->quota_per_period = LLONG_MAX;
  else if (sscanf(quota_buf, "%lld", &constraint->quota_per_period) != 1)
    return UV_EINVAL; // conversion failed
```

santigimeno reviewed (approved):
LGTM with a couple of comments.

waltoss commented:
@santigimeno @bnoordhuis any idea when this can be merged ?

bnoordhuis commented:
I'll go ahead and merge it. Thanks for the pull request.
